src/execution_tree/compiler/compiler.cpp

```diff
@@ -386,37 +375,38 @@ namespace phylanx { namespace execution_tree { namespace compiler
                 external_function(f, sequence_number, default_locality_));
 
             static std::string define_function_("define-function");
-            sequence_number =
+            name_parts.primitive = define_function_;
+            name_parts.sequence_number =
                 snippets_.sequence_numbers_[define_function_]++;
 
-            // create define_function helper object
-            std::string define_function_name =
-                std::to_string(sequence_number) + "$" + full_name;
-
-            f = primitive_function{default_locality_}(
-                    define_function_ + "$" + define_function_name,
-                    name_);
+            f = primitive_function{default_locality_}(name_parts, name_);
 
             // set the body for the compiled function
-            primitive_operand(f.arg_, define_function_name, name_).set_body(
-                hpx::launch::sync, std::move(handle_lambda(args, body).arg_));
+            primitive_operand(f.arg_, compose_primitive_name(name_parts), name_)
+                .set_body(hpx::launch::sync,
+                    std::move(handle_lambda(args, body).arg_));
 
             // define-function shouldn't return a function that evaluates
             // to itself, let it return nil{} instead
-            return always_nil{}(std::move(define_function_name));
+            return always_nil{}(std::move(name_parts));
         }
 
         function handle_variable_reference(std::string name,
             ast::expression const& expr)
         {
             ast::tagged id = ast::detail::tagged_id(expr);
+            primitive_name_parts name_parts(name, -1, id.id, id.col);
```

Comment:
This is not good as the `tagged_id` could be 'empty'

Reply:
If it's empty it will contain -1 and doesn't mess with anything else, but saves an `if`.

Reply:
Ok, fair point.